[Done] Separating Missing (NaG) and Empty

caspervdw opened the pull request:
Closes #28, closes #29

This is some work separating the concepts of missing geometries (not-a-geometry or NaG) values and empty geometries. The idea is to keep a single NaG object (like None) that has its GEOSGeometry pointer set to NULL.

All geometry-returning functions propagate the NaG; predicates return False for any import NaG; integer returning functions return -1 for NaG.

I added two additional functions to identify NaG values: `is_geometry` and `is_null`.

Further there should be functions that create arrays of empty geometries.

jorisvandenbossche commented:
Cool! I will try to take a look at it one of the coming days. Currently at EuroScipy conference, so a bit busy, but, I will sit together here with one of the numpy developers who is interested in seeing the dtype stuff :-)

jorisvandenbossche commented on src/pygeom.c:
```diff
@@ -9,14 +9,20 @@
-        Py_INCREF(Geom_Empty);
-        return (PyObject *) Geom_Empty;
+        if (NaG == NULL) {
+            /* First time: initialize the NaG object */
```
Shouldn't we do this first time initialization in the ufuncs module initialization function?

jorisvandenbossche commented on src/ufuncs.c:
```diff
@@ -502,24 +485,36 @@ static void Y_i_func(char **args, npy_intp *dimensions,
+            /* NaG results in -1 */
+            *(npy_int *)op1 = -1;
+        } else {
+            /* trust that GEOS calls the error handler if necessary */
```
This is mentioned many times, but wondering, do we actually test this? (in which occasion would GEOS raise an error? sometimes one invalid geometries?)

caspervdw replied:
There are some cases tested:
https://github.com/caspervdw/pygeos/search?q=pygeos.GEOSException&unscoped_q=pygeos.GEOSException

GEOS should do a callback to the error message handler (`HandleGEOSError`), which I defined such that a GEOSException is set. I have never encountered the situation where GEOS didn't do this.

jorisvandenbossche commented on src/ufuncs.c:
```diff
@@ -169,26 +157,34 @@ static void YY_b_func(char **args, npy_intp *dimensions,
+            /* call the GEOS function */
+            ret = func(context_handle, in1, in2);
+            /* return for illegal values (trust HandleGEOSError for SetErr) */
+            if ((ret < 0) | (ret > 1)) { return; }
```
`ret != 1` instead ?

jorisvandenbossche commented on pygeos/predicates.py:
```diff
@@ -73,18 +75,77 @@ def is_empty(geometry, **kwargs):
+    >>> is_null(None)
+    True
+    >>> is_null(float("nan"))
+    True
```
So this is somewhat related to the comment I made below about limiting the allowed input (checking each element from being float gives an overhead, and for me it is not necessary a needed feature to be able to pass all of three potential missing-like objects)

caspervdw commented on pygeos/set_operations.py:
```diff
@@ -78,8 +76,8 @@ def intersection_all(geometries, axis=0, **kwargs):
-    >>> intersection_all([[line_1, line_2, Empty]], axis=1).tolist()
-    [<pygeos.Empty>]
+    >>> intersection_all([[line_1, line_2, NaG]], axis=1).tolist()
+    [<pygeos.NaG>]
```
Yes, I'd like do continue that discussion in #37

caspervdw commented on pygeos/measurement.py:
```diff
@@ -39,7 +39,7 @@ def distance(a, b):
     1.0
     >>> distance(Geometry("POLYGON ((3 0, 5 0, 5 5, 3 5, 3 0))"), point)
     3.0
-    >>> distance(Empty, point)
```
GEOS actuall returns `0.0` if you do this. Which is strange. We could change the GEOS API here.

jorisvandenbossche replied:
It's indeed strange.In GeoPandas we return NaN (and I kept that in the recent refactoring, as we had been doing that for a long time). See the bullet point in https://github.com/geopandas/geopandas/pull/1062#issuecomment-520127793. But, it's also annoying to need to deviate from GEOS ..

caspervdw replied:
I did pushed the change, I think it makes no sense at all to return 0.0 when computing the distance to an empty set.

caspervdw commented on pygeos/predicates.py:
```diff
@@ -73,18 +75,77 @@ def is_empty(geometry, **kwargs):
+    True
+    >>> is_geometry(None)
+    False
+    >>> is_geometry(float("nan"))
```
Agreed, all these None types are really not necessary. I dropped the NaG and now only None is supported.

caspervdw commented:
Thanks for the remarks, I made some changes:
 - Dropped the NaG and NaN support in favour of only None
 - is_null is now replaced by is_missing (which is just np.equals(arr, None) under the hood)

caspervdw commented:
> Additionally to those, it would alos be useful to have a fast ufunc exposed to just check that the input only contains "valid objects" (in pygeos array's context, so a geometry or a NaG) as well. So basically a is_geometry(a) | is_null(a), but as single ufunc this can be more efficient.

Maybe good to wait to see how this numpy-dtype extension thing plays out?

jorisvandenbossche commented:
> Maybe good to wait to see how this numpy-dtype extension thing plays out?

On the short term, it unfortunately won't play out. See the discussion in #34. It will need changes in numpy to be feasible (but changes they would welcome / want to plan).

jorisvandenbossche commented on pygeos/constructive.py:
```diff
@@ -51,8 +51,8 @@ def boundary(geometry, **kwargs):
     <pygeos.Geometry LINESTRING (0 0, 1 0, 1 1, 0 1, 0 0)>
-    >>> boundary(Geometry("MULTIPOINT (0 0, 1 2)"))
-    <pygeos.NaG>
+    >>> boundary(Geometry("MULTIPOINT (0 0, 1 2)")) is None
```
Alternative could be to do `print(..)` which then actually shows `None` (but both are of course workarounds for None having no repr)

caspervdw commented:
> > Maybe good to wait to see how this numpy-dtype extension thing plays out?
> 
> On the short term, it unfortunately won't play out. See the discussion in #34. It will need changes in numpy to be feasible (but changes they would welcome / want to plan).

Oh I wasn 't following the discussion. So I implemented 3 ufuncs `is_missing`, `is_geometry`, `is_valid_input`
